benchmark/main.cpp

```diff
@@ -56,979 +50,1350 @@
+        if (ret < 0 || static_cast<size_t>(ret) >= sizeof(title)) {
+            mbedtls_printf("Failed to compose title string using "
+                           "mbedtls_snprintf(): %d\n", ret);
+            goto exit;
+        }
+
+        memset(buf, 0, sizeof(buf));
+        memset(tmp, 0, sizeof(tmp));
+
+        ret = mbedtls_ccm_setkey(&ccm, MBEDTLS_CIPHER_ID_AES, tmp, keysize);
+        if (ret != 0) {
+            PRINT_ERROR(ret, "mbedtls_gcm_setkey()");
+            goto exit;
         }
+
+        BENCHMARK_FUNC_CALL(title,
+                            mbedtls_ccm_encrypt_and_tag(&ccm, BUFSIZE, tmp, 12,
+                                    NULL, 0, buf, buf, tmp,
+                                    16));
     }
-#endif
-#if defined(MBEDTLS_CMAC_C)
-    if( todo->aes_cmac )
-    {
-        unsigned char output[16];
-        const mbedtls_cipher_info_t *cipher_info;
-        mbedtls_cipher_type_t cipher_type;
-        int keysize;
-
-        cipher_type = MBEDTLS_CIPHER_AES_128_ECB;
-        for( keysize = 128; keysize <= 256; keysize += 64 )
-        {
-            mbedtls_snprintf( title, sizeof( title ), "AES-CMAC-%d", keysize );
-
-            memset( buf, 0, sizeof( buf ) );
-            memset( tmp, 0, sizeof( tmp ) );
-
-            cipher_info = mbedtls_cipher_info_from_type( cipher_type );
-
-            TIME_AND_TSC( title,
-                          mbedtls_cipher_cmac( cipher_info, tmp, keysize,
-                                               buf, BUFSIZE, output ) );
-            cipher_type = (mbedtls_cipher_type_t)( cipher_type + 1 );
+
+exit:
+    mbedtls_ccm_free(&ccm);
+
+    return ret;
+}
+#endif /* MBEDTLS_AES_C && MBEDTLS_CCM_C */
+
+#if defined(MBEDTLS_AES_C) && defined(MBEDTLS_CMAC_C)
+MBED_NOINLINE static int benchmark_aes_cmac()
+{
+    int ret = 0;
+    unsigned char output[16];
+    const mbedtls_cipher_info_t *cipher_info;
+    mbedtls_cipher_type_t cipher_type;
+    int keysize;
+
+    cipher_type = MBEDTLS_CIPHER_AES_128_ECB;
+    for (keysize = 128; keysize <= 256; keysize += 64) {
+        ret = mbedtls_snprintf(title, sizeof(title), "AES-CMAC-%d", keysize);
+        if (ret < 0 || static_cast<size_t>(ret) >= sizeof(title)) {
+            mbedtls_printf("Failed to compose title string using "
+                           "mbedtls_snprintf(): %d\n", ret);
+            goto exit;
         }
 
-        memset( buf, 0, sizeof( buf ) );
-        memset( tmp, 0, sizeof( tmp ) );
-        TIME_AND_TSC( "AES-CMAC-PRF-128",
-                      mbedtls_aes_cmac_prf_128( tmp, 16, buf, BUFSIZE,
-                                                output ) );
+        memset(buf, 0, sizeof(buf));
+        memset(tmp, 0, sizeof(tmp));
+
+        cipher_info = mbedtls_cipher_info_from_type(cipher_type);
+        if (cipher_info == NULL) {
+            mbedtls_printf("mbedtls_cipher_info_from_type() returned NULL\n");
+            goto exit;
+        }
+
+        BENCHMARK_FUNC_CALL(title,
```

Comment:
CMAC could also return `MBEDTLS_AES_FEATURE_NOT_SUPPORTED` in case no alternative CMAC.
This is problematic because of the MACRO though

Reply:
Is this still a problem?

Reply:
Unfortunately, the macro does make it a bit awkward to add the extra checks. I did add [this](https://github.com/ARMmbed/mbed-os-example-tls/pull/194/files#diff-11d19f91de60e62a1952ab3e8b07cd0cR160) line of code to check for MBEDTLS_AES_FEATURE_NOT_SUPPORTED on calls like this.

I agree that the code is not ideal, but my main motivation for this refactor was to patch up the reliability issues. So I tried to minimize the development time by no changing all these macros. Perhaps we should consider replacing them when it is needed to check more feature not supported errors.